File: src/audit_reports/document_origin.py
"""Compare a fresh official download with acquired bytes; never replace either.

Matching bytes establish source revision agreement, not text/table accuracy.
Different revisions and unavailable URLs remain explicit review outcomes.
"""
from __future__ import annotations

from datetime import datetime, timezone
import hashlib
from pathlib import Path

import fitz

from .document_acquisition import fetch_source, unwrap_pdf
from .document_corpus import Filing
from .document_corpus_resume import metadata
from .document_corpus_store import PREFIX, _error_code, _json
from .document_quality import source_identity_review


def _sha(body):
    return hashlib.sha256(body).hexdigest()


def observe_origin(store, filing: Filing, acquisition_key: str | None, url: str, patterns: dict, *,
                   reviewed_member: dict | None = None, fetch=fetch_source, checked_at: str | None = None):
    result = {'schema_version': 'document-origin-review-1', 'filing': filing.as_dict(), 'source_url': url,
              'acquisition_key': acquisition_key, 'requested_archive_selection': reviewed_member,
              'checked_at': checked_at or datetime.now(timezone.utc).isoformat(),
              'acquisition': None, 'transport': None, 'origin_pdf': None,
              'semantically_verified': False,
              'engine': {'pymupdf': fitz.VersionBind, 'implementation': {
                  name: _sha(Path(__file__).with_name(name).read_bytes()) for name in
                  ('document_origin.py', 'document_acquisition.py', 'document_quality.py')}}}
    acquired = None
    if acquisition_key:
        try:
            response = store.client.get_object(Bucket=store.bucket, Key=acquisition_key)
        except Exception as error:
            if _error_code(error) not in ('404', 'NoSuchKey'):
                raise
        else:
            try:
                acquired = response['Body'].read()
            finally:
                response['Body'].close()
            if len(acquired) != response['ContentLength']:
                raise ValueError('Acquisition read was truncated during origin comparison')
            result['acquisition'] = {'key': acquisition_key, 'sha256': _sha(acquired), 'bytes': len(acquired),
                                     'version': metadata(acquisition_key, response)}
    artifacts = {}
    try:
        transport, response = fetch(url)
    except Exception as error:
        result.update(status='origin_unavailable', error=str(error))
        return result, artifacts
    result.update(response=response, transport={'sha256': _sha(transport), 'bytes': len(transport)})
    artifacts['transport'] = transport
    try:
        body, selection = unwrap_pdf(transport, reviewed_member)
        result['selection'] = selection
        result['origin_pdf'] = {'sha256': _sha(body), 'bytes': len(body)}
        artifacts['origin_pdf'] = body
        leading = []
        with fitz.open(stream=body, filetype='pdf') as pdf:
            if pdf.needs_pass or not len(pdf):
                raise ValueError('Origin PDF requires a password or has no pages')
            for number in range(min(3, len(pdf))):
                raw = pdf[number].get_text('dict', flags=fitz.TEXTFLAGS_DICT & ~fitz.TEXT_PRESERVE_IMAGES,
                                           clip=fitz.INFINITE_RECT())
                spans = []
                for block_id, block in enumerate(raw['blocks']):
                    if block['type'] != 0:
                        continue
                    for line_id, line in enumerate(block['lines']):
                        for span in line['spans']:
                            spans.append({'id': len(spans), 'block': block_id, 'line': line_id,
                                          'text': span['text'], 'bbox': list(span['bbox'])})
                leading.append({'page': number + 1, 'spans': spans})
            result['page_count'] = len(pdf)
        result['origin_leading_pages'] = leading
        result['origin_identity'] = source_identity_review(filing, leading, patterns)
        result['related_pdf_content_capture'] = 'pending' if selection.get('unselected_pdf_members') else 'not_applicable'
    except Exception as error:
        result.update(status='origin_needs_review', error=str(error))
        return result, artifacts
    if acquired is None:
        status = 'acquisition_missing'
    elif acquired == body:
        status = 'matches_acquired_bytes'
    else:
        status = 'different_pdf_revision'
        # Only an observed serialized wrapper may be normalized. Metadata,
        # punctuation, PDF objects and visible contents are never discarded.
        if acquired.startswith(b'\xac\xed\x00\x05'):
            try:
                normalized, wrapper = unwrap_pdf(acquired)
            except ValueError as error:
                result['acquisition_wrapper_error'] = str(error)
            else:
                result['acquisition_wrapper'] = wrapper
                if normalized == body:
                    status = 'same_pdf_after_acquisition_wrapper'
    result['status'] = status
    return result, artifacts


def publish_origin(store, result: dict, artifacts: dict[str, bytes], patterns: dict) -> dict:
    """Keep downloaded evidence and an immutable review before updating its index."""
    import json

    filing = Filing(**result['filing'])
    if result.get('schema_version') != 'document-origin-review-1' or result.get('semantically_verified') is not False:
        raise ValueError('Invalid source-origin review')
    observed_at = datetime.fromisoformat(result['checked_at'])
    if observed_at.tzinfo is None or observed_at.utcoffset().total_seconds() != 0:
        raise ValueError('Origin observation time must have an explicit UTC offset')
    acquired = result['acquisition']
    if acquired:
        current = metadata(acquired['key'], store.client.head_object(Bucket=store.bucket, Key=acquired['key']))
        if current != acquired['version']:
            raise ValueError('Acquisition changed during origin review')
    expected = {name for name in ('transport', 'origin_pdf') if result.get(name) is not None}
    if set(artifacts) != expected:
        raise ValueError('Origin review is missing its downloaded evidence')
    def retained_download(_url):
        if 'transport' not in artifacts:
            raise RuntimeError(result['error'])
        return artifacts['transport'], result['response']
    checked, retained = observe_origin(store, filing, result['acquisition_key'], result['source_url'], patterns,
        reviewed_member=result['requested_archive_selection'], fetch=retained_download, checked_at=result['checked_at'])
    if checked != result or retained != artifacts:
        raise ValueError('Origin review differs from acquired bytes and retained transport')
    value = dict(result)
    for name, body in artifacts.items():
        record = result[name]
        if _sha(body) != record['sha256'] or len(body) != record['bytes']:
            raise ValueError('Origin artifact differs from its observed bytes')
        key = (f"{PREFIX}transports/{record['sha256']}/original.bin" if name == 'transport' else
               f"{PREFIX}sources/{record['sha256']}/original.pdf")
        store._immutable(key, body, 'application/octet-stream' if name == 'transport' else 'application/pdf')
        value[name] = {**record, 'key': key}
    base = f'{PREFIX}origins/{filing.bank_ticker}/{filing.period}/{filing.kind}/'
    payload = _json(value)
    digest = _sha(payload)
    key = base + digest + '.json'
    store._immutable(key, payload, 'application/json')
    revision = {'key': key, 'sha256': digest, 'bytes': len(payload), 'checked_at': value['checked_at'],
                'status': value['status'], 'acquisition_sha256': acquired['sha256'] if acquired else None}
    for _ in range(8):
        previous, etag = store._read(base + 'index.json')
        index = json.loads(previous) if previous else {'schema_version': 'document-origin-index-1',
            'filing': filing.as_dict(), 'current': None, 'revisions': [], 'semantically_verified': False}
        if (index['filing'] != filing.as_dict() or index['schema_version'] != 'document-origin-index-1'
                or index.get('semantically_verified') is not False):
            raise ValueError('Origin index differs from its filing')
        if revision not in index['revisions']:
            index['revisions'].append(revision)
        index['current'] = max(index['revisions'], key=lambda r: (r['checked_at'], r['sha256']))
        body = _json(index)
        if body == previous:
            return {**value, 'review_key': key, 'index_key': base + 'index.json'}
        try:
            store.client.put_object(Bucket=store.bucket, Key=base + 'index.json', Body=body,
                                    ContentType='application/json', **({'IfMatch': etag} if etag else {'IfNoneMatch': '*'}))
            readback, _ = store._read(base + 'index.json')
            retained_index = json.loads(readback)
            if (retained_index.get('filing') != filing.as_dict()
                    or retained_index.get('schema_version') != 'document-origin-index-1'
                    or retained_index.get('semantically_verified') is not False
                    or revision not in retained_index.get('revisions', [])):
                raise ValueError('Origin index readback does not retain this observation')
            return {**value, 'review_key': key, 'index_key': base + 'index.json'}
        except Exception as error:
            if _error_code(error) not in ('412', 'PreconditionFailed'):
                raise
    raise RuntimeError('Origin index changed concurrently; retry this review')

File: web/app/lib/document-origin.ts
/** Independent official-source observations; byte agreement never approves contents. */
import { CORPUS_PREFIX, type CorpusBucket, type FilingIdentity } from "./document-corpus";
import { readRecoveryArtifact } from "./document-recovery";

type Artifact = { key: string; sha256: string; bytes: number };
export type OriginReview = {
  schema_version: "document-origin-review-1"; filing: FilingIdentity; checked_at: string;
  status: "matches_acquired_bytes" | "same_pdf_after_acquisition_wrapper" | "different_pdf_revision"
    | "acquisition_missing" | "origin_unavailable" | "origin_needs_review";
  source_url: string; semantically_verified: false; error?: string;
  acquisition: { sha256: string; bytes: number } | null;
  transport: Artifact | null; origin_pdf: Artifact | null;
  origin_identity?: { status: string }; related_pdf_content_capture?: string;
  selection?: { unselected_pdf_members?: { name: string; bytes: number; sha256: string }[] };
};
const record = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);
const hash = (v: unknown): v is string => typeof v === "string" && /^[a-f0-9]{64}$/.test(v);
const count = (v: unknown): v is number => typeof v === "number" && Number.isSafeInteger(v) && v > 0;
const sameFiling = (v: unknown, f: FilingIdentity) => record(v) && Object.entries(f).every(([k, value]) => v[k] === value);
const time = (v: unknown): v is string => typeof v === "string" && /(?:Z|\+00:00)$/.test(v) && Number.isFinite(Date.parse(v));
const statuses = new Set(["matches_acquired_bytes", "same_pdf_after_acquisition_wrapper", "different_pdf_revision",
  "acquisition_missing", "origin_unavailable", "origin_needs_review"]);

export async function getOriginReview(bucket: CorpusBucket, filing: FilingIdentity): Promise<OriginReview | null> {
  const base = `${CORPUS_PREFIX}origins/${filing.bank_ticker}/${filing.period}/${filing.kind}/`;
  const object = await bucket.get(base + "index.json");
  if (!object) return null;
  if (object.size > 4_000_000) throw new Error("Oversized origin index");
  const index: unknown = await object.json();
  if (!record(index) || index.schema_version !== "document-origin-index-1" || !sameFiling(index.filing, filing)
      || index.semantically_verified !== false || !Array.isArray(index.revisions) || !record(index.current)) {
    throw new Error("Invalid origin index binding");
  }
  const current = index.current;
  if (!hash(current.sha256) || current.key !== `${base}${current.sha256}.json` || !count(current.bytes)
      || !time(current.checked_at) || !statuses.has(String(current.status))
      || !index.revisions.some(r => record(r) && ["key", "sha256", "bytes", "checked_at", "status", "acquisition_sha256"]
        .every(k => r[k] === current[k]))) throw new Error("Invalid origin revision");
  const bytes = await readRecoveryArtifact(bucket, current as Artifact);
  const value: unknown = JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes));
  if (!record(value) || value.schema_version !== "document-origin-review-1" || !sameFiling(value.filing, filing)
      || value.semantically_verified !== false || value.status !== current.status || value.checked_at !== current.checked_at
      || typeof value.source_url !== "string" || !/^https?:\/\//.test(value.source_url)
      || value.error !== undefined && typeof value.error !== "string") throw new Error("Origin receipt binding mismatch");
  if (value.acquisition !== null && (!record(value.acquisition) || !hash(value.acquisition.sha256)
      || !count(value.acquisition.bytes))) throw new Error("Invalid acquired-source observation");
  if ((record(value.acquisition) ? value.acquisition.sha256 : null) !== current.acquisition_sha256) {
    throw new Error("Origin acquisition differs from its index");
  }
  for (const name of ["transport", "origin_pdf"]) {
    const entry = value[name];
    if (entry === null) continue;
    if (!record(entry) || !hash(entry.sha256) || !count(entry.bytes)
        || entry.key !== (name === "transport" ? `${CORPUS_PREFIX}transports/${entry.sha256}/original.bin`
          : `${CORPUS_PREFIX}sources/${entry.sha256}/original.pdf`)) throw new Error("Invalid origin artifact key");
  }
  if (value.origin_identity !== undefined && (!record(value.origin_identity) || typeof value.origin_identity.status !== "string")) {
    throw new Error("Invalid origin identity observation");
  }
  if (value.selection !== undefined) {
    if (!record(value.selection)) throw new Error("Invalid origin archive selection");
    const members = value.selection.unselected_pdf_members;
    if (members !== undefined && (!Array.isArray(members) || members.some(m => !record(m)
        || typeof m.name !== "string" || !count(m.bytes) || !hash(m.sha256)))) throw new Error("Invalid related PDF members");
  }
  const acquiredHash = record(value.acquisition) ? value.acquisition.sha256 : null;
  const originHash = record(value.origin_pdf) ? value.origin_pdf.sha256 : null;
  if (value.status === "matches_acquired_bytes" && (!acquiredHash || acquiredHash !== originHash)
      || value.status === "different_pdf_revision" && (!acquiredHash || !originHash || acquiredHash === originHash)
      || value.status === "acquisition_missing" && (acquiredHash !== null || !originHash)
      || value.status === "origin_unavailable" && (value.transport !== null || value.origin_pdf !== null)
      || value.status === "same_pdf_after_acquisition_wrapper" && (!acquiredHash || !originHash || !record(value.acquisition_wrapper))) {
    throw new Error("Origin comparison contradicts observed source hashes");
  }
  return value as unknown as OriginReview;
}
